x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,449 @@
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package s3
+
+import (
+	"bufio"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+	"time"
+
+	awssdk "github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/defaults"
+	"github.com/aws/aws-sdk-go-v2/aws/external"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/s3iface"
+	"github.com/aws/aws-sdk-go-v2/service/sqs"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/filebeat/channel"
+	"github.com/elastic/beats/filebeat/harvester"
+	"github.com/elastic/beats/filebeat/input"
+	"github.com/elastic/beats/filebeat/util"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+const inputName = "s3"
+
+var (
+	// The maximum number of messages to return. Amazon SQS never returns more messages
+	// than this value (however, fewer messages might be returned).
+	maxNumberOfMessage int64 = 10
+
+	// The duration (in seconds) for which the call waits for a message to arrive
+	// in the queue before returning. If a message is available, the call returns
+	// sooner than WaitTimeSeconds. If no messages are available and the wait time
+	// expires, the call returns successfully with an empty list of messages.
+	waitTimeSecond int64 = 10
+)
+
+func init() {
+	err := input.Register(inputName, NewInput)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Input is a input for s3
+type Input struct {
+	outlet    channel.Outleter
+	config    config
+	awsConfig awssdk.Config
+	logger    *logp.Logger
+	close     chan struct{}
+}
+
+type s3Info struct {
+	name   string
+	key    string
+	region string
+	arn    string
+}
+
+type sqsMessage struct {
+	Records []struct {
+		EventSource string `json:"eventSource"`
+		AwsRegion   string `json:"awsRegion"`
+		EventName   string `json:"eventName"`
+		S3          struct {
+			Bucket struct {
+				Name string `json:"name"`
+				Arn  string `json:"arn"`
+			} `json:"bucket"`
+			Object struct {
+				Key string `json:"key"`
+			} `json:"object"`
+		} `json:"s3"`
+	} `json:"Records"`
+}
+
+// NewInput creates a new s3 input
+func NewInput(cfg *common.Config, outletFactory channel.Connector, context input.Context) (input.Input, error) {
+	cfgwarn.Beta("s3 input type is used")
+
+	logger := logp.NewLogger(inputName)
+
+	config := defaultConfig
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrap(err, "failed unpacking config")
+	}
+
+	outlet, err := outletFactory(cfg, context.DynamicFields)
+	if err != nil {
+		return nil, err
+	}
+
+	err = config.validate()
+	if err != nil {
+		return nil, errors.Wrapf(err, "validation for s3 input config failed: config = %v", config)
+	}
```

Comment:
`"validation for s3 input config failed: config = %v", config)`
I think the above  could leaks the credentials in the logs?

Reply:
If you implement `Validate() error` with a capital **V** that method will be called when you unpack the values so you should not need to call it manually.

Reply:
good to know! Thanks!! I'm gonna change it to `Validate()`!